src/Microsoft.ML.Data/DataView/Transposer.cs

```diff
@@ -704,13 +704,16 @@ private void EnsureValid()
                 Utils.EnsureSize(ref _cbuff, vecLen);
                 for (int s = 0; s < vecLen; ++s)
                 {
-                    var temp = new VBuffer<T>(_len, _counts[s], _values[s], _indices[s]);
-                    if (temp.Count < _len / 2)
+                    int count = _counts[s];
+                    T[] values = _values[s];
+                    int[] indices = _indices[s];
+                    var temp = new VBuffer<T>(_len, count, values, indices);
+                    if (count < _len / 2)
                     {
                         // Already sparse enough, I guess. Swap out the arrays.
                         Utils.Swap(ref temp, ref _cbuff[s]);
-                        _indices[s] = temp.Indices ?? new int[_len];
```

Comment:
@TomFinley - it looks like the `Microsoft.ML.Runtime.RunTests.TestPredictors.FastTreeBinaryClassificationCategoricalSplitTest` test is failing because of this change. Any ideas how to make this work given out plan of not directly exposing the `VBuffer.Indices` and `VBuffer.Values` arrays?

One thought I had was to change 

```
            // The transposed contents of _colStored.
            private VBuffer<T>[] _cbuff; // Working intermediate column-wise buffer.
```
To instead hold a new structure that had references to `VBuffer`, `T[]` and `int[]`. That way we could still access the arrays that back the `VBuffer` that we created. I'm super open to any other ideas though.

Reply:
Are the values different in some way, or are you just sparsifying differently?

If the latter, which is my belief, then there must be a bug inside FastTree. It is probably treating dense or sparse input vectors differently in a way that it should not. Perhaps @codemzs knows something of this.

---
In reply to: [232116206](https://github.com/dotnet/machinelearning/pull/1580#discussion_r232116206) [](ancestors = 232116206)

Reply:
@tomfinley - I "fixed" this in the latest commit.  Please take a look and let me know what you think.

---
In reply to: [232424930](https://github.com/dotnet/machinelearning/pull/1580#discussion_r232424930) [](ancestors = 232424930,232116206)